core/src/server/rpc_module.rs

```diff
@@ -816,98 +846,196 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
-			Ok(())
-		} else {
-			Err(SubscriptionAcceptRejectError::RemotePeerAborted)
-		}
+	/// Get the method name.
+	pub fn method_name(&self) -> &str {
+		self.method
 	}
 
-	/// Return the subscription ID if the the subscription was accepted.
+	/// Send out a response on the subscription and wait until there is capacity.
 	///
-	/// [`SubscriptionSink::accept`] should be called prior to this method.
-	pub fn subscription_id(&self) -> Option<RpcSubscriptionId<'static>> {
-		if self.id.is_some() {
-			// Subscription was not accepted.
-			None
-		} else {
-			Some(self.uniq_sub.sub_id.clone())
+	///
+	/// Returns
+	/// - `Ok(())` if the message could be sent.
+	/// - `Err(err)` if the connection or subscription was closed.
+	///
+	/// # Cancel safety
+	///
+	/// This method is cancel-safe and dropping a future loses its spot in the waiting queue.
+	pub async fn send(&self, msg: SubscriptionMessage) -> Result<(), DisconnectError> {
+		// Only possible to trigger when the connection is dropped.
+		if self.is_closed() {
+			return Err(DisconnectError(msg));
 		}
+
+		self.inner.send(msg.0).await.map_err(Into::into)
 	}
 
-	/// Send a message back to subscribers.
+	/// Attempts to immediately send out the message as JSON string to the subscribers but fails if the
+	/// channel is full or the connection/subscription is closed
 	///
-	/// Returns
-	/// - `Ok(true)` if the message could be send.
-	/// - `Ok(false)` if the sink was closed (either because the subscription was closed or the connection was terminated),
-	/// or the subscription could not be accepted.
-	/// - `Err(err)` if the message could not be serialized.
-	pub fn send<T: Serialize>(&mut self, result: &T) -> Result<bool, serde_json::Error> {
-		// Cannot accept the subscription.
-		if let Err(SubscriptionAcceptRejectError::RemotePeerAborted) = self.accept() {
-			return Ok(false);
+	///
+	/// This differs from [`SubscriptionSink::send`] where it will until there is capacity
+	/// in the channel.
+	pub fn try_send(&mut self, msg: SubscriptionMessage) -> Result<(), TrySendError> {
+		// Only possible to trigger when the connection is dropped.
+		if self.is_closed() {
+			return Err(TrySendError::Closed(msg));
 		}
 
-		self.send_without_accept(result)
+		self.inner.try_send(msg.0).map_err(Into::into)
+	}
+
+	/// Returns whether the subscription is closed.
+	pub fn is_closed(&self) -> bool {
+		self.inner.is_closed() || !self.is_active_subscription()
+	}
+
+	/// Completes when the subscription has been closed.
+	pub async fn closed(&self) {
+		// Both are cancel-safe thus ok to use select here.
+		tokio::select! {
+			_ = self.inner.closed() => (),
+			_ = self.unsubscribe.unsubscribed() => (),
+		}
+	}
+
+	/// Build message that will be serialized as JSON-RPC notification.
+	///
+	/// You need to call this method prior to send out a subscription notification.
+	pub fn build_message<T: Serialize>(&self, result: &T) -> Result<SubscriptionMessage, serde_json::Error> {
```

Comment:
Did you decide not to have a separate `Message::build` type thing in the end?

Reply:
It was just super noisy to implement, lemme give it one more try.